### ISSUE TITLE
Electrum 3.1.3

### ISSUE THREAD

**Issue by Libre12** [hashcat/hashcat#1806]
Hi guys,

I forgot my wallet password, running Electrum 3.1.3.

So Hashcat still doesn't support this wallet format?

Anyway to work around it?

Thanks,
Libre

**Comment by philsmd** [comment]
did you try running electrum2john.py ?

does the output start with $electrum$2* or is it $electrum$3* ?

Hashcat supports salt type 1 and 2 currently, not sure if your type is not supported (3 ?)

**Comment by Libre12** [comment]
where do i find electrum2john.py ?

I have downloaded and compiled hashcat directly from github on my Ubuntu but I can't find this file.

**Comment by s3inlc** [comment]
Like the name is indicating, electrum2john.py is part of John the Ripper, you can find it in its repository: https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/run/electrum2john.py

**Comment by Libre12** [comment]
> Like the name is indicating, electrum2john.py is part of John the Ripper, you can find it in its repository: https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/run/electrum2john.py

Thanks

**Comment by Libre12** [comment]
> did you try running electrum2john.py ?
> 
> does the output start with $electrum$2* or is it $electrum$3* ?
> 
> Hashcat supports salt type 1 and 2 currently, not sure if your type is not supported (3 ?)

$electrum$5*

Hashcat supports this version?

**Comment by jsteube** [comment]
$electrum$5* is not supported

**Comment by teasider** [comment]
> did you try running electrum2john.py ?
> 
> does the output start with $electrum$2* or is it $electrum$3* ?
> 
> Hashcat supports salt type 1 and 2 currently, not sure if your type is not supported (3 ?)

Sorry for hijacking, but when using the found hash (for example $electrum#2*XXXXXXX) - do i need to include $electrum$2*XXXXXXX as a whole in the hash file or just whatever comes after (just the XXXXXXXX, the hash itself)

$electrum$2*XXXXXXXXXXXXXXXXXXXXXX
or
XXXXXXXXXXXXXXXXXXXXXXXX

thanks in advance!

**Comment by philsmd** [comment]
I just found out that version 5 of the format is a little hack: https://github.com/magnumripper/JohnTheRipper/blob/1070c8bb38c8fbf5a75e5c49341d98631bcabeff/run/electrum2john.py#L50 and a special case for the electrum 2.8 wallet format.

electrum2john.py should have given you the warning "Forcing generation of truncated hash, this is not tested well", can you confirm that you got this warning @Libre12 ?

BTW: I found some list of example formats on this github link: https://github.com/spesmilo/electrum/blob/94c4cb44d723133bb18dfea578d266f2e36c9cb6/electrum/tests/test_storage_upgrade.py (but they seem to be all unencrypted, so not that useful). Didn't try them yet which work with hashcat, but it could be a handy link to see what is supported and what needs to be changed etc

**Comment by philsmd** [comment]
I think that looking at the jtr source code for $electrum$5* hashes (https://github.com/magnumripper/JohnTheRipper/blob/1b4f0cc3a44ecb9512a70e0bc01b2a39a1882e9f/src/electrum_fmt_plug.c#L367-L394), the addition of zlib to hashcat's dependencies (https://github.com/hashcat/hashcat/commit/0c8768c520dd9fcac88f9a22238c57239205db28) makes this $electrum$5* feature request at least a little bit more feasible.
The problem is that we still need a hook for performing those operations (inflate the byte stream,  i.e. decompress) and therefore it's very different from $electrum$1* or $electrum$2* hashes.
Furthermore, looking at the code of john the ripper, we would also need elliptic curve operations (secp256k1), for which we also would need the code (libraries etc). It's actually quite a huge difference to the other electrum formats (with need of hooks, decompression and ecc)

**Comment by maxya** [comment]
@philsmd 
so.. how do I proceed with this ? 
```
 ./electrum2john.py my_wallet
Forcing generation of truncated hash, this is not tested well!
my_wallet:$electrum$5*
```

**Comment by philsmd** [comment]
The natural order would be to first implement $electrum$3* , then $electrum$4* and finally $electrum$5*.

If I'm not totally mistaken we would need the hook and elliptic curve cryptography for both the $electrum$4 and $electrum$5 hashes. in theory (for the purpose of getting $electrum$5 hashes working soon) you could skip the $electrum$3, because it's more closer to $electrum$2.... but $electrum$4 and $electrum$5 need almost the same code (except decompression for $electrum$5 which we already support with deps/zlib/ ).

So how do we proceed @maxya , I would suggest that you start implementing the hook code and look for elliptic curve code that we could use (compatible license etc). As a starter, you could first get started with an (hopefully) easier hash mode: $electrum$3 and after you are accustomed with the hashcat code and especially the module feature, you can start experiencing with hooks (see -m 11600 = 7-Zip for an example of hook and decompression).

We could also try to ping @mcovalt which seem to have helped with other/older ethereum salt-types see https://github.com/hashcat/hashcat/pull/1805 . 
Thank you !

**Comment by philsmd** [comment]
I've now implemented $electrum$3 support with https://github.com/hashcat/hashcat/pull/2021 . yeah, this one was really easy, as predicted.

So the missing electrum hash support is now reduced to $electrum$4 and $electrum$5 ... anyone wants to try to implement one of these (they are way more difficult, that's for sure ;) ) ?

**Comment by philsmd** [comment]
I've done some little reseach again and it seems that many projects use this MIT licensed (very good for us !!!) elliptic curve with parameters Secp256k1 (https://www.secg.org/sec2-v2.pdf) used by bitcoin:

https://github.com/bitcoin-core/secp256k1/tree/master/src

and

https://github.com/bitcoin-core/secp256k1/blob/84a808598b3aa70847d73a0ce43f7893e302fd90/COPYING#L1

Therefore, if we add these secp256k1 source code into the deps/ (e.g. deps/secp256k1) folder, we have both the zlib and elliptic curve cryptography (ecc) code for the hook needed for $electrum$4 and $electrum$5 , that's my educated guess right now. Maybe there are more obstacles that I didn't consider yet, but besides including that specific elliptic curve and adding the new kernels with a hook function, I do not see much more trouble for adding this new hash type (I think we can do both $electrum$4 and $electrum$5 with just one new hash type... but that's still to be confirmed).
Thanks

**update**: I see quite little interest in this and my "discoveries", so it's currently quite low in my priorities to look into this. Please still mention it here if you are looking into implementing it s.t. we are avoid doing it in parallel.

**Comment by philsmd** [comment]
I've spend some bit of (free) time again to check the details of the salt type 5 electrum algorithm and I came up with a single file proof of concept (POC) perl script that checks the password for the jtr example hash from https://github.com/magnumripper/JohnTheRipper/blob/c41c6f042a0644105572e916abcc9a51eed144a1/src/electrum_fmt_plug.c#L94:

```
#!/usr/bin/env perl

# Author: philsmd
# Date: June 2019
# License: public domain, credits go to philsmd and hashcat, props also to jtr for their electrum_fmt_plug.c code

use strict;
use warnings;

use Crypt::OpenSSL::EC;
use Crypt::OpenSSL::Bignum::CTX;

use Crypt::CBC;
use Crypt::PBKDF2;

use Digest::SHA qw (sha512);

use Compress::Zlib;

#
# Example data
#

# $electrum$5*0328e536dd1fbbb85d78de1a8c21215d4646cd87d6b6545afcfb203e5bb32e0de4*61b1e287a5acff4b40e4abd73ff62dc233c1c7a6a54b3270949281b9d44bc6e746743733360500718826e50bb28ea99a6378dc0b0c578e9d0bf09c667671c82a1bd71c8121edbb4c9cbca93ab0e17e218558ead81755e62b0d4ad547aa1b3beb0b9ee43b11270261c9b38502f00e7f6f096811b7fdae6f3dce85c278d3751fec044027054218ccf20d404bab24380b303f094704e626348a218f44ab88ce2ac5fa7d450069fca3bb53f9359dbbaad0ea1b3859129b19c93ed7888130f8a534f84a629c67edc150a1c5882a83cb0add4615bb569e8dc471de4d38fc8b1e0b9b28040b5ea86093fcdeceaedb6b8f073f6f0ee5541f473a4b1c2bfae4fc91e4bbb40fa2185ecfa4c72010bcf8df05b1a7db45f64307dbc439f8389f0e368e38960b6d61ac88c07ce95a4b03d6d8b13f4c7dc7d7c447097865235ab621aeef38dc4172bf2dc52e701132480127be375fe98834f16d9895dce7f6cdfe900a2ce57eaa6c3036c1b9a661c3c9adbf84f4adfe6d4d9fa9f829f2957cfb353917dc77fd8dd4872b7d90cb71b7d3a29c9bfe3440e02449220acba410fa0af030f51aa2438f7478dbb277d62613112e4eebc66d5d7bdba793fb2073d449954f563284819189ffb5dbcdeb6c95c64bc24e0ef986bce07bafe96ab449ae2b6edaf4f98ffbd392a57bd93c2359444ec4046ae65b440adb96b6e4eef9d06bb04d2f3fa2e4175165bcadbf7e13cc3b6e65e67df901f96a2f154bc763b56b3736a335e1d1bc16e99736f757a4ae56c099645c917360b1ecf8dcefc7281541c6ff65d87cadab4a48f1f6b7b73a3e5a67e2e032abb56b499e73a9f3b69ce065e43b0174639785ae30635d105ebcc827dcf9b19bdd1a92879a5d4bc4e12b5630c188b1b96e3c586e19901b8f96084bcd59b2f4b201a3a8b6e633a5c194901d4609add9671b0bcc12b2b94ae873d201258b36315484e4b9c5f5d6289656baa93eec9e92aec88e2d73d86b9e3d1f24294e3d8ebe9a9f2f6edfbf28f530670c5b086fc4f74df89b4e4cbe06ee7e45cbd238b599d19c2d5da5523b12b1e7050ea0a9b47a5d22c6c3fc476f814f9705dc7ed3aeb1b44fc6b4d69f02a74963dce5057c3c049f92e595a4da5035cffc303a4cb162803aa3f816527a7e466b8424789a0d77e26819615662420c370457e29fcc1938fd754f3acfd21416ce3ab27e9febbc0e24fc7055eddc31e48faa014f9f3695c2e956f0e6c94c507a8d2f8c3aeb4b98b69b6340b6a3acb1acdde9581279f78ee10687616360c018e9f67d6c8bb5950e8fdabd3d0d5808824975aa4a50f88581472212f24ad58a700fe4787642b973924575fe71d1ecd7b2b6acd363f48c40bdd55f35f60a06dee544c266e608fd5a6d263f745e8b11d1160638eb301adfd1a88eddf6d0ccb9e1021e0bde9cf5163583a202b3dc95c255c8cc24*ec90c1ff54632e7c8cfb812eeb14d7ec49ddaf576dca10bfb16f965e6106ce48:btcr-test-password

my $ephemeral_pubkey = pack ("H*", "0328e536dd1fbbb85d78de1a8c21215d4646cd87d6b6545afcfb203e5bb32e0de4");

my $data_buf = pack ("H*", "61b1e287a5acff4b40e4abd73ff62dc233c1c7a6a54b3270949281b9d44bc6e746743733360500718826e50bb28ea99a6378dc0b0c578e9d0bf09c667671c82a1bd71c8121edbb4c9cbca93ab0e17e218558ead81755e62b0d4ad547aa1b3beb0b9ee43b11270261c9b38502f00e7f6f096811b7fdae6f3dce85c278d3751fec044027054218ccf20d404bab24380b303f094704e626348a218f44ab88ce2ac5fa7d450069fca3bb53f9359dbbaad0ea1b3859129b19c93ed7888130f8a534f84a629c67edc150a1c5882a83cb0add4615bb569e8dc471de4d38fc8b1e0b9b28040b5ea86093fcdeceaedb6b8f073f6f0ee5541f473a4b1c2bfae4fc91e4bbb40fa2185ecfa4c72010bcf8df05b1a7db45f64307dbc439f8389f0e368e38960b6d61ac88c07ce95a4b03d6d8b13f4c7dc7d7c447097865235ab621aeef38dc4172bf2dc52e701132480127be375fe98834f16d9895dce7f6cdfe900a2ce57eaa6c3036c1b9a661c3c9adbf84f4adfe6d4d9fa9f829f2957cfb353917dc77fd8dd4872b7d90cb71b7d3a29c9bfe3440e02449220acba410fa0af030f51aa2438f7478dbb277d62613112e4eebc66d5d7bdba793fb2073d449954f563284819189ffb5dbcdeb6c95c64bc24e0ef986bce07bafe96ab449ae2b6edaf4f98ffbd392a57bd93c2359444ec4046ae65b440adb96b6e4eef9d06bb04d2f3fa2e4175165bcadbf7e13cc3b6e65e67df901f96a2f154bc763b56b3736a335e1d1bc16e99736f757a4ae56c099645c917360b1ecf8dcefc7281541c6ff65d87cadab4a48f1f6b7b73a3e5a67e2e032abb56b499e73a9f3b69ce065e43b0174639785ae30635d105ebcc827dcf9b19bdd1a92879a5d4bc4e12b5630c188b1b96e3c586e19901b8f96084bcd59b2f4b201a3a8b6e633a5c194901d4609add9671b0bcc12b2b94ae873d201258b36315484e4b9c5f5d6289656baa93eec9e92aec88e2d73d86b9e3d1f24294e3d8ebe9a9f2f6edfbf28f530670c5b086fc4f74df89b4e4cbe06ee7e45cbd238b599d19c2d5da5523b12b1e7050ea0a9b47a5d22c6c3fc476f814f9705dc7ed3aeb1b44fc6b4d69f02a74963dce5057c3c049f92e595a4da5035cffc303a4cb162803aa3f816527a7e466b8424789a0d77e26819615662420c370457e29fcc1938fd754f3acfd21416ce3ab27e9febbc0e24fc7055eddc31e48faa014f9f3695c2e956f0e6c94c507a8d2f8c3aeb4b98b69b6340b6a3acb1acdde9581279f78ee10687616360c018e9f67d6c8bb5950e8fdabd3d0d5808824975aa4a50f88581472212f24ad58a700fe4787642b973924575fe71d1ecd7b2b6acd363f48c40bdd55f35f60a06dee544c266e608fd5a6d263f745e8b11d1160638eb301adfd1a88eddf6d0ccb9e1021e0bde9cf5163583a202b3dc95c255c8cc24");

my $password = "btcr-test-password";


#
# Start
#

# key derivation from password (private key for secp256k1 elliptic curve operations)

my $pbkdf2 = Crypt::PBKDF2->new
(
  hasher     => Crypt::PBKDF2->hasher_from_algorithm ('HMACSHA2', 512),
  iterations => 1024,
  output_len => 64
);

my $private_key = $pbkdf2->PBKDF2 ("", $password);

my $method = Crypt::OpenSSL::EC::EC_GFp_simple_method (); # or Crypt::OpenSSL::EC::EC_GFp_mont_method ()

my $group = Crypt::OpenSSL::EC::EC_GROUP::new ($method);


# secp256k1 elliptic curve parameters

my $p = Crypt::OpenSSL::Bignum->new_from_hex ("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f");
my $a = Crypt::OpenSSL::Bignum->new_from_hex ("0000000000000000000000000000000000000000000000000000000000000000");
my $b = Crypt::OpenSSL::Bignum->new_from_hex ("0000000000000000000000000000000000000000000000000000000000000007");

my $ctx = Crypt::OpenSSL::Bignum::CTX->new ();

Crypt::OpenSSL::EC::EC_GROUP::set_curve_GFp ($group, $p, $a, $b, $ctx);

my $Gx = Crypt::OpenSSL::Bignum->new_from_hex ("79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798");
my $Gy = Crypt::OpenSSL::Bignum->new_from_hex ("483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8");

my $G = Crypt::OpenSSL::EC::EC_POINT::new ($group);

Crypt::OpenSSL::EC::EC_POINT::set_affine_coordinates_GFp ($group, $G, $Gx, $Gy, $ctx);

my $order    = Crypt::OpenSSL::Bignum->new_from_hex ("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141");
my $cofactor = Crypt::OpenSSL::Bignum->new_from_hex ("0000000000000000000000000000000000000000000000000000000000000001");

Crypt::OpenSSL::EC::EC_GROUP::set_generator ($group, $G, $order, $cofactor); # or cofactor = Crypt::OpenSSL::Bignum->one ()


# scalar

# hash mod GROUP_ORDER (example: f6a35ba782a2c832efdfb1f497a397a9edf2de48f13fb259d4f5e1f30c3e2641)

my $m = Crypt::OpenSSL::Bignum->new_from_hex (unpack ("H*", $private_key));


# point (public key, ephemeral_pubkey)

# example:
# my $Qx = Crypt::OpenSSL::Bignum->new_from_hex ("28e536dd1fbbb85d78de1a8c21215d4646cd87d6b6545afcfb203e5bb32e0de4");
# my $Qy = Crypt::OpenSSL::Bignum->new_from_hex ("31649e46497ea56e9de83abef6a91669a7f64ce6a7fa24e50aa4739cf81fcd71");
#Crypt::OpenSSL::EC::EC_POINT::set_affine_coordinates_GFp ($group, $Q, $Qx, $Qy, $ctx);

my $Q = Crypt::OpenSSL::EC::EC_POINT::new ($group);

Crypt::OpenSSL::EC::EC_POINT::oct2point ($group, $Q, $ephemeral_pubkey, $ctx);


# multiply

my $result = Crypt::OpenSSL::EC::EC_POINT::new ($group);

my $n = Crypt::OpenSSL::Bignum->zero ();

Crypt::OpenSSL::EC::EC_POINT::mul ($group, $result, $n, $Q, $m, $ctx);


# debug example/code of getting the x/y coordinates

# my $result_x = Crypt::OpenSSL::Bignum->new ();
# my $result_y = Crypt::OpenSSL::Bignum->new ();
#
# Crypt::OpenSSL::EC::EC_POINT::get_affine_coordinates_GFp ($group, $result, $result_x, $result_y, $ctx);
#
# print $result_x->to_hex () . "/" . $result_y->to_hex () . "\n";


# get compressed public/shared key format

my $public_key = Crypt::OpenSSL::EC::EC_POINT::point2oct ($group, $result, &Crypt::OpenSSL::EC::POINT_CONVERSION_COMPRESSED, $ctx);


# hash the compressed public key with sha512 ()

my $key_data = sha512 ($public_key);


# decrypt the data with AES128

my $iv      = substr ($key_data,  0, 16);
my $aes_key = substr ($key_data, 16, 16);

my $aes = Crypt::CBC->new ({
  cipher      => "Crypt::Rijndael",
  keysize     => 16,
  literal_key => 1,
  header      => "none",
  iv          => $iv,
  key         => $aes_key
});

my $decrypted_data = $aes->decrypt ($data_buf);


# some early reject/validation steps:

# first test:

if (substr ($decrypted_data, 0, 2) ne "\x78\x9c")
{
  print STDERR "ERROR: Couldn't AES decrypt the data correctly. Wrong password?\n";

  exit (1);
}

# second test:

if ((ord (substr ($decrypted_data, 2, 1)) & 0x07) != 0x05)
{
  print STDERR "ERROR: Couldn't AES decrypt the data correctly. Wrong password?\n";

  exit (1);
}


# decompress/inflate:

my $inflator = inflateInit (-WindowBits => MAX_WBITS);

my $status;

my $decompressed_data = "";

($decompressed_data, $status) = $inflator->inflate ($decrypted_data);


# final validation of data:

if (length ($status) > 0)
{
  print STDERR "ERROR: Couldn't decompress the data. Wrong password?\n";

  exit (1);
}

if ((substr ($decompressed_data, 0, 7) ne "{\n    \"") && (substr ($decompressed_data, 0, 8) ne "{\r\n    \""))
{
  print STDERR "ERROR: decompressed data is not a valid wallet file. Wrong password?\n";

  exit (1);
}


# if we reached this code section we are done and have found the correct password:

print "password found: '$password'\n";
```

It took me quite some time to get all the details correct (elliptic curve setup and multiply operations, AES decryption, zlib inflate etc), but this is very important stuff, because we could use this directly in the perl test modules (tools/test_modules/) for test.sh/test.pl tests ... but also now it's very easy to debug and see what needs to be done with this minimal script.

**Comment by philsmd** [comment]
update: I created a git branch to try to add support for Electrum 4/5 with libsecp256k1 and zlib (as explained above): https://github.com/philsmd/hashcat/commits/electrum_4_and_5

It's working for me and especially with the support of multi-threading in the CPU hook function (https://github.com/hashcat/hashcat/commit/a8555fa048517bf352c0ba16745d467f6c2ed837) it is quite effective/fast.

There are some things to still work on, like making the src/Makefile automatically do all the stuff (like gmplib / deps/secp256k1 compilation) and making sure it works on all supported platforms macOS/linux/windows/cross-compilation.

I tested the cross-compilation and an executable was generated (first baby steps)... would need to test if this works on windows too.
